[SPARK-15515] [SQL] Error Handling in Running SQL Directly On Files

gatorsmile opened the pull request:
#### What changes were proposed in this pull request?

This PR is to address the following issues: 
- **ISSUE 1:** For ORC source format, we are reporting the strange error message when we did not enable Hive support: 

``` SQL
SQL Example: 
  select id from `org.apache.spark.sql.hive.orc`.`file_path`
Error Message:
  Table or view not found: `org.apache.spark.sql.hive.orc`.`file_path`
```

Instead, we should issue the error message like:

```
Expected Error Message:
   The ORC data source must be used with Hive support enabled
```
- **ISSUE 2:** For the Avro format, we report the strange error message like:

The example query is like

``` SQL
SQL Example: 
  select id from `avro`.`file_path`
  select id from `com.databricks.spark.avro`.`file_path`
Error Message:
  Table or view not found: `com.databricks.spark.avro`.`file_path`
```

The desired message should be like:

```
Expected Error Message:
  Failed to find data source: avro. Please use Spark package http://spark-packages.org/package/databricks/spark-avro"
```
- ~~**ISSUE 3:** Unable to detect incompatibility libraries for Spark 2.0 in Data Source Resolution. We report a strange error message:~~

**Update**: The latest code changes contains 
- For JDBC format, we added an extra checking in the rule `ResolveRelations` of `Analyzer`. Without the PR, Spark will return the error message like: `Option 'url' not specified`. Now, we are reporting `Unsupported data source type for direct query on files: jdbc` 
- Make data source format name case incensitive so that error handling behaves consistent with the normal cases. 
- Added the test cases for all the supported formats. 
#### How was this patch tested?

Added test cases to cover all the above issues

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
SPARK-11633

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into joinMakeCopy

gatorspark committed:
Merge remote-tracking branch 'upstream/master'

gatorspark committed:
Merge branch 'master' of https://github.com/gatorsmile/spark

Commit:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
converge

gatorsmile committed:
converge

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
fix.

SparkQA commented:
**[Test build #59229 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59229/consoleFull)** for PR 13283 at commit [`948340b`](https://github.com/apache/spark/commit/948340b15e504effcbc160deeb2e859a31714f39).

SparkQA commented:
**[Test build #59229 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59229/consoleFull)** for PR 13283 at commit [`948340b`](https://github.com/apache/spark/commit/948340b15e504effcbc160deeb2e859a31714f39).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59229/
Test PASSed.

gatorsmile committed:
code clean

SparkQA commented:
**[Test build #59249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59249/consoleFull)** for PR 13283 at commit [`f6a4517`](https://github.com/apache/spark/commit/f6a45177609c3c2ab7cd376dd6eb1a4527e693e7).

gatorsmile committed:
better output

SparkQA commented:
**[Test build #59251 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59251/consoleFull)** for PR 13283 at commit [`26d72ae`](https://github.com/apache/spark/commit/26d72ae697b57fb8ad1a4f0357af0b726a76e443).

SparkQA commented:
**[Test build #59251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59251/consoleFull)** for PR 13283 at commit [`26d72ae`](https://github.com/apache/spark/commit/26d72ae697b57fb8ad1a4f0357af0b726a76e443).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ClassNotFoundException(CapturedException):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59251/
Test FAILed.

gatorsmile committed:
style fix.

SparkQA commented:
**[Test build #59254 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59254/consoleFull)** for PR 13283 at commit [`c313ffc`](https://github.com/apache/spark/commit/c313ffc88488229c45e74cde97210c8686e1c745).

gatorsmile commented:
Just realized this PR is related to https://github.com/apache/spark/pull/13201. Also tried to improve the exception reporting in Python through hiding the long java stack trace. 

@zsxwing @marmbrus Could you please help me review the changes? Thanks!

SparkQA commented:
**[Test build #59254 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59254/consoleFull)** for PR 13283 at commit [`c313ffc`](https://github.com/apache/spark/commit/c313ffc88488229c45e74cde97210c8686e1c745).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59254/
Test PASSed.

SparkQA commented:
**[Test build #59249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59249/consoleFull)** for PR 13283 at commit [`f6a4517`](https://github.com/apache/spark/commit/f6a45177609c3c2ab7cd376dd6eb1a4527e693e7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59249/
Test PASSed.

rxin commented:
Just want to say --- this is a really cool change!

(I didn't actually look at the details. Would be great for somebody else to do a detailed review.)

gatorsmile commented:
Thank you!

zsxwing commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:
```diff
@@ -132,12 +131,11 @@ case class DataSource(
-                "Please check if your library is compatible with Spark 2.0", e)
+          case e: NoClassDefFoundError =>
+            // e.getMessage is the class name of provider2. Instead, we use provider here.
+            if (spark2RemovedClasses.contains(provider)) {
```
In my previous PR, I want to provide a better message for e.g., `org.apache.spark.sql.DataFrame` not found. It usually happens when calling some method (has a `org.apache.spark.sql.DataFrame` parameter) in a class that is compiled with an old Spark. Obviously, here `provider` won't be `org.apache.spark.sql.DataFrame`.

zsxwing replied:
You should not change this. If `provider` is not found, `loadClass` will throw `ClassNotFoundException`. If a class used by provider are not found, `NoClassDefFoundError` will be thrown.

gatorsmile replied:
Will revert the changes for `NoClassDefFoundError`. Thanks!

zsxwing commented:
> ISSUE 3: Unable to detect incompatibility libraries for Spark 2.0 in Data Source Resolution. We report a strange error message:
> SQL Example:
>   select id from `org.apache.spark.sql.sources.HadoopFsRelationProvider`.`file_path`
> Error Message: 
>   Table or view not found: `org.apache.spark.sql.sources.HadoopFsRelationProvider`.`file_path`

This is not an issue you need to fix. `HadoopFsRelationProvider` is just an interface in Spark 1.6. The user should not use it like this. If someone sees HadoopFsRelationProvider is not found, it's usually a link issue. E.g., `com.databricks.spark.avro.DefaultSource` extends `HadoopFsRelationProvider`, however, `HadoopFsRelationProvider` has been removed in 2.0, so when loading `com.databricks.spark.avro.DefaultSource`, it will throw `NoClassDefFoundError(HadoopFsRelationProvider)` instead of `ClassNotFoundException(com.databricks.spark.avro.DefaultSource)`

zsxwing commented:
@gatorsmile I saw it reports `org.apache.spark.sql.AnalysisException: Failed to find data source: mydatabase. Please find packages at http://spark-packages.org;; line 1 pos 15` for the following statement: `sql("select id from `mydatabase`.`file_path`")`. I'm not familiar with the table name resolution. Is it correct?

zsxwing commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:
```diff
@@ -108,21 +108,20 @@ case class DataSource(
-                val className = error.getMessage
-                if (spark2RemovedClasses.contains(className)) {
-                  throw new ClassNotFoundException(s"$className is removed in Spark 2.0. " +
+                // error.getMessage is the class name of provider2. Instead, we use provider here.
```
In a second thought, I don't think we need this `if` branch. Could you just remove it?

zsxwing replied:
This is for link issues. But it will be `NoClassDefFoundError`

gatorsmile replied:
Sure, will do it. Thanks!

zsxwing commented on python/pyspark/sql/utils.py:
```diff
@@ -77,6 +83,8 @@ def deco(*a, **kw):
                 raise QueryExecutionException(s.split(': ', 1)[1], stackTrace)
             if s.startswith('java.lang.IllegalArgumentException: '):
                 raise IllegalArgumentException(s.split(': ', 1)[1], stackTrace)
+            if s.startswith('java.lang.NoClassDefFoundError: '):
```
The Python changes are not necessary. Right?

gatorsmile replied:
I think it is still helpful. This is to hide the long java stack trace for improving the error reporting. It was introduced in the PR https://github.com/apache/spark/pull/7135.

zsxwing replied:
`NoClassDefFoundError` rarely happens unless we miss something in `spark2RemovedClasses`. Hiding it makes us hard to figure out the real place quickly.

gatorsmile replied:
uh, I see. Will revert it back. Thanks!

gatorsmile committed:
address comments.

SparkQA commented:
**[Test build #59296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59296/consoleFull)** for PR 13283 at commit [`3ac2b93`](https://github.com/apache/spark/commit/3ac2b930a1629a2fd21602a7ceb22429fb8ce372).

gatorsmile commented:
It sounds like we need to verify all the possible source types we can support. Let me add them. Thanks!

SparkQA commented:
**[Test build #59296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59296/consoleFull)** for PR 13283 at commit [`3ac2b93`](https://github.com/apache/spark/commit/3ac2b930a1629a2fd21602a7ceb22429fb8ce372).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59296/
Test PASSed.

gatorsmile committed:
address comments.

SparkQA commented:
**[Test build #59332 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59332/consoleFull)** for PR 13283 at commit [`76f4f80`](https://github.com/apache/spark/commit/76f4f80f962e0271a2073a4cb8de0d513013cf87).

gatorsmile commented:
**Update**: The latest code changes contains 
- For JDBC format, we added an extra checking in the rule `ResolveRelations` of `Analyzer`. Without the PR, Spark will return the error message like: `Option 'url' not specified`. Now, we are reporting `Unsupported data source type for direct query on files: jdbc` 
- Make data source format name case incensitive so that error handling behaves consistent with the normal cases. 
- Added the test cases for all the supported formats.

SparkQA commented:
**[Test build #59332 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59332/consoleFull)** for PR 13283 at commit [`76f4f80`](https://github.com/apache/spark/commit/76f4f80f962e0271a2073a4cb8de0d513013cf87).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59332/
Test FAILed.

gatorsmile committed:
fix test case.

SparkQA commented:
**[Test build #59346 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59346/consoleFull)** for PR 13283 at commit [`b9e12f8`](https://github.com/apache/spark/commit/b9e12f8742e76984445f9d498248704b1c9e9973).

SparkQA commented:
**[Test build #59346 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59346/consoleFull)** for PR 13283 at commit [`b9e12f8`](https://github.com/apache/spark/commit/b9e12f8742e76984445f9d498248704b1c9e9973).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59346/
Test PASSed.

gatorsmile commented:
@zsxwing Now, code is ready for review. Thanks!

zsxwing commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -448,6 +448,14 @@ class Analyzer(
+     * Data source formats that were not supported in direct query on file
+     */
+    private final val unsupportedFileQuerySource = Set(
+      "org.apache.spark.sql.jdbc",
```
Not sure about this. Looks hard to maintain. E.g., you forget to add `JdbcRelationProvider`. @rxin What do you think?

marmbrus replied:
Can we just check to see if it extends `FileFormat`?

gatorsmile replied:
@zsxwing @marmbrus Agree. It is hard to maintain the list. First, will add the missing class names to the list.

To implement what @marmbrus suggested, I think we need to do the following:

 - Users' input are not full class names. To dynamically load the class, we still need to have a list like:
https://github.com/apache/spark/blob/361ebc282b2d09dc6dcf21419a53c5c617b1b6bd/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L80-L108
To avoid duplicating the codes, we need to move this list to a place both `Analyzer` and `DataSource` can access. Any suggestion here?

 - Dynamically loading the class and check whether the class extends `FileFormat`. We need to do something similar like https://github.com/apache/spark/blob/361ebc282b2d09dc6dcf21419a53c5c617b1b6bd/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L128-L132

 - To avoid duplicate the codes, when unable to loading the class, we still follow the existing way to handle it. That is, returns the `UnresolvedRelation` `u`. 

Is my understanding right?

Thank you for your suggestions!

marmbrus replied:
Couldn't the logic go in `ResolveDataSource`?

gatorsmile replied:
: ) Yeah. Let me try it. Thank you!

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into runSQLAgainstFile

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge branch 'runSQLAgainstFile' into runSQLAgainstFileNew

gatorsmile committed:
address comments

SparkQA commented:
**[Test build #59554 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59554/consoleFull)** for PR 13283 at commit [`e5c08f2`](https://github.com/apache/spark/commit/e5c08f2a54c2b622f273625fe32a9fd369735500).

SparkQA commented:
**[Test build #59554 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59554/consoleFull)** for PR 13283 at commit [`e5c08f2`](https://github.com/apache/spark/commit/e5c08f2a54c2b622f273625fe32a9fd369735500).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59554/
Test PASSed.

gatorsmile commented:
@zsxwing @marmbrus How about the latest code changes? Let me know if anything I can further improve. Thanks!

zsxwing commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:
```diff
@@ -38,6 +38,10 @@ private[sql] class ResolveDataSource(sparkSession: SparkSession) extends Rule[Lo
           sparkSession,
           paths = u.tableIdentifier.table :: Nil,
           className = u.tableIdentifier.database.get)
+        if (dataSource.isFileFormat() == Option(false)) {
```
`isFileFormat` is not necessary. You can use `providingClass` like this:

```Scala
        val notSupportDirectQuery = try {
          !classOf[FileFormat].isAssignableFrom(dataSource.providingClass)
        } catch {
          case NonFatal(e) => false
        }
        if (notSupportDirectQuery) {
          throw new AnalysisException("Unsupported data source type for direct query on files: " +
            s"${u.tableIdentifier.database.get}")
        }
```

gatorsmile replied:
Thank you very much! Will use your version! It is much better. Thanks again!

gatorsmile committed:
address comments.

SparkQA commented:
**[Test build #59699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59699/consoleFull)** for PR 13283 at commit [`9fae469`](https://github.com/apache/spark/commit/9fae46981667e38ee801a06d79d4afac77ed4372).

SparkQA commented:
**[Test build #59699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59699/consoleFull)** for PR 13283 at commit [`9fae469`](https://github.com/apache/spark/commit/9fae46981667e38ee801a06d79d4afac77ed4372).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59699/
Test PASSed.

gatorsmile commented:
@zsxwing @marmbrus Any remaining issue? Thanks!

zsxwing commented:
LGTM. Merging to master and 2.0. Thanks!

tedyu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:
```diff
@@ -38,6 +40,16 @@ private[sql] class ResolveDataSource(sparkSession: SparkSession) extends Rule[Lo
+        val notSupportDirectQuery = try {
+          !classOf[FileFormat].isAssignableFrom(dataSource.providingClass)
+        } catch {
+          case NonFatal(e) => false
```
When would this happen ?

Should true be returned here ?

zsxwing replied:
@tedyu If people use `select * from db_name.table_name`, it will throw an exception. Still need to continue for such cases.

tedyu replied:
Thanks Ryan.
